docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -487,24 +486,28 @@ spec:
 
 ##### Cluster upgrade (using create-swap-and-delete)
 
-- Triggered on any changes to Version or the InfrastructureTemplate used (the KubeadmConfiguration is treated and validated as immutable).
+- Triggered on any changes to Version or the InfrastructureTemplate
+  - The KubeadmConfiguration, for the first iteration, is treated and validated as immutable and changes to it do not trigger an upgrade
   - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
-    - Possible implementations include (not necessary to hammer out exact implementation right now):
-      - Manually bumping a Generation field on the KubeadmControlPlane object
-      - Adding a field that is not the Generation field for users to manually bump to trigger an upgrade
-      - Adding a timestamp field, something like `nextUpgrade` which can be set to a specific time in the future (far future if no upgrades are desired) and an upgrade will run when that timestamp is passed. Good for scheduling upgrades/SLOs & quickly modifying to NOW or yesterday if an upgrade is required immediately. Would potentially update the field value to some interval (2 months from now?)
-- An upgrade will look like this:
-  - Serially go through each control plane replica not at the current config hash
-    - Check health of any replicas for the current config hash
-    - Uniquely mark the existing replica machine through setting a label and persisting the change (cluster.x-k8s.io/control-plane/selected-for-upgrade)
-    - Check cluster health
-    - Provision a new machine at the correct version
-    - Poll on health of replicas with the current config hash
-    - Poll on cluster health with the correct replica number
-    - Remove the replica’s etcd member from the etcd cluster
-    - Update the kubeadm generated config map
-    - Delete the marked controlplane machine
-- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).
+      - By adding a timestamp field, `upgradeAfter`, which can be set to a specific time in the future
+        - Set to far future if no upgrades are desired
+        - An upgrade will run when that timestamp is passed
```

Comment:
The controller will attempt to upgrade any Machines that have a CreationTimestamp that is older than `upgradeAfter`

Reply:
This doesn't make sense to me, can you provide an example?

If a machine was created on February 21st, 2020 and has an upgradeAfter value of February 23rd, 2020, the controller would try to upgrade that?

Reply:
upgradeAfter is a filed on the KubeadmControlPlane

Reply:
that doesn't change the example. If the KubeadmControlPlane controller upgrades any control plane machine that was created before the `upgradeAfter` field, it seems to me it would create a new machine over and over until actual time has passed the `upgradeAfter` field and therefore the creationTimestamp has surpassed KubeadmControlPlane.

I may be thinking about this backwards  🤔

Reply:
After further discussion, we do need to revisit the behavior related to the current reconciliation of the `upgradeAfter` field.